Duotone: Limit SVG filter output to used filters

jeryj opened the pull request:
Rework of https://github.com/WordPress/gutenberg/pull/48374
Fixes https://github.com/WordPress/gutenberg/issues/38299.

## What?
This PR only adds the needed duotone code to the frontend output if that duotone definition is being used.

## Why?
There are currently eight globally defined duotone styles, and all of their CSS and SVG definitions are output on the frontend of a WordPress website, even if no images are using the duotone filters. The base raw HTML from TT3 on the Sherbert theme variation results in a 76.4kb request. Without any SVG or duotone-related code, the same page HTML is 67.4kb, which means there is up to 9kb of unnecessary duotone-related HTML on every request.

![raw html view of unnecessary SVG code in html](https://user-images.githubusercontent.com/967608/225126237-249a0f42-ff88-4d02-99be-e4b675cdcb2c.png)


## How?
There's quite a bit of changes required to make this happen. We've gone with a route that adds a new `WP_Duotone` class that handles what filters need to be added to the page based on what duotone filters, if any, are in use. Here's a general overview of the high-level changes being made, and why:

### 1. Remove all duotone output from global styles
These are handled by removing the actions related to SVG filters:

```
remove_action( 'wp_body_open', 'wp_global_styles_render_svg_filters' );
remove_action( 'in_admin_header', 'wp_global_styles_render_svg_filters' );
```

### 2. Register all duotone filters from global styles and store in the WP_Duotone class
 
`WP_Duotone::$global_styles_presets` is an array that scrapes all of the global styles duotone filters. These could be defined from theme.json globally, per theme, or in custom global styles. We need to do this to have quick access to the duotone filter definitions if we come across a block that needs to access these styles.

### 3. Register all block-level duotone filters from global styles and store in the WP_Duotone class

`WP_Duotone::$global_styles_block_names` is an array that scrapes all of the blocks that use a duotone filter from global styles, and stores the slug of the duotone filter it uses, such as `default-filter`, `blue-orange`, etc. We need to do this in order to decide if a block being rendered has a duotone filter we need to apply to it.

### 4. Process all blocks to determine the correct duotone output needed

This is the same as before, but we're only outputting the duotone code if needed. On the 'render_block' filter, we process every block and determine:
- Is there a duotone filter we need to apply on this block
- If not, return early
- If we do, decide which duotone filter we need and how to build it
- Add the duotone info we need to `WP_Duotone::$output`

### 5. Output the CSS
On the `'wp_enqueue_scripts' action, output all the necessary CSS for duotone filters collected from `WP_Duotone::$output`.

### 6. Output the SVG
On the `'wp_footer' action, output all necessary SVG for duotone filters from `WP_Duotone::$output`.


<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

## Testing Instructions
These are the different types of Duotone that need testing:
1. Set the duotone in Global Styles to all blocks of a certain type.
2. Set the duotone on a per blocks basis.

For each of these it possible to set the duotone to a `preset` value, a custom value, or `unset`.
For presets, it is possible to set them using two different formats: `var:preset|duotone|blue-orange` or `var(--wp--preset--duotone--blue-orange)`.

It is also important to test that _only_ the SVG filters and CSS that is needed on the page is output - rather than outputting all the presets even if they aren't used.

To test an individual block:
1. Add an image to a post
2. Click on the "filter" option in the block toolbar
3. Select a Duotone option
4. Save the post
5. Open the post in the frontend and check that the correct Duotone filter has been applied

To test Global Styles
1. Open the Global Styles interface in the Site Editor
2. Open "Blocks"
3. Find the image block
4. Click on "filter"
5. Select a Duotone filter
6. Save the Global Styles settings
7. Open the site in the frontend and check that all image blocks have the filter applied (unless they specify their own filters)

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

jeryj committed:
Adds WP_Duotone class to set-up preset and theme-defined blocks using duotone presets

jeryj committed:
Fix incorrect filter name

jeryj committed:
Check if duotone filter is applied on a block in theme.json

This doesn't do anything at the moment. On block render, it checks if the block has duotone filters set via theme.json/global styles/custom styles and returns early if not.

jeryj committed:
Rename duotone class static variables and functions

jeryj committed:
Set WP_Duotone:: array and only output those via wp_footer

jeryj committed:
Only output used duotone presets

Co-authored-by: scruffian <ben@escruffian.com>

jeryj committed:
Cleanup, removal, and only output used SVG filters and CSS

jeryj committed:
Support var:preset|duotone|default-filter syntax for block name styles

jeryj committed:
Refactor to handle both duotone pipe and css var style syntax for duotone preset

jeryj committed:
Renamed output_presets to output, as it also includes custom dutone output. Name could still be better.

jeryj committed:
Fix theme json vs block styles duotone order of application

Refactored out duotone specific code from theme json gutenberg globla styles. Switched theme_json_register_declarations from action to filter.

jeryj committed:
Renamed filter to be more accurate, but should refactor out

Doesn't seem like a legitimately useful filter at the moment. Only for us to be able to use global styles in our duotone-specific output.

jeryj committed:
Simplify getting global styles presets

We were already getting the presets via the global settings, so the filter was unnecessary.

jeryj committed:
Undo theme json class duotone refactoring and remove filter

The filter isn't necesary for us to access what we need. I've removed the refactor as it would be a larger change that can be done in a different PR if needed.

jeryj committed:
Remove unused code

ajlende commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -144,7 +144,7 @@ class WP_Theme_JSON_Gutenberg {
+			'value_func'        => null, // Don't output CSS Custom Properties for duotone.
 			'css_vars'          => '--wp--preset--duotone--$slug',
 			'classes'           => array(),
 			'properties'        => array( 'filter' ),
```
I'm pretty sure the entire duotone section of `PRESETS_METADATA` can be removed since generating the preset is done by duotone.php now.

ajlende replied:
Actually, it looks like the constructor still uses it for keying presets by origin which we need later. That being said, I still think we can remove the `css_vars` key and have that solely defined in duotone.php as part of the cleanup.

Also, I noticed that the theme.json `styles.blocks['core/block'].filter.duotone` filters are being applied on the frontend, but not the editor, so there are also likely changes that need to be done in the block-editor duotone.js hook and use-global-styles-output.js as well.

ajlende replied:
The `properties` key is also probably redundant now since we're no longer generating properties here.

jeryj committed:
Rename save_ to set_ class method names

jeryj committed:
Fix duotone test to use real global preset value

jeryj committed:
Tests for WP_Duotone::gutenberg_get_slug_from_attr()

jeryj committed:
Rework gutenberg_get_slug_from_attr to use regex for more accurate matches

jeryj committed:
Tests for WP_Duotone::is_preset

jeryj committed:
Rework is_preset to also check if the preset exists

jeryj committed:
Remove array_key_exists checks for presets within the block filter

ajlende commented on phpunit/block-supports/duotone-test.php:
```diff
@@ -37,4 +37,41 @@ public function test_gutenberg_render_duotone_support_custom() {
+		return array(
+			'pipe-slug'   => array( 'var:preset|duotone|blue-orange', 'blue-orange' ),
+			'css-var'   => array( 'var(--wp--preset--duotone--blue-orange)', 'blue-orange' ),
+			'css-var-weird-chars'   => array( 'var(--wp--preset--duotone--.)', '.' ),
```
`.` isn't a valid identifier token according to the [CSS spec](https://w3c.github.io/csswg-drafts/css-syntax-3/#ident-token-diagram). 

CSS custom properties always take the first `--` branch of this railroad diagram and then allow for `[a-zA-Z0-9_-]` and [a set of non-ascii code points](https://w3c.github.io/csswg-drafts/css-syntax-3/#non-ascii-ident-code-point).

![ident-token railroad diagram](https://user-images.githubusercontent.com/5129775/225076073-eb4b3ad8-63be-4b90-ada7-1244bbbaaf23.png)

It looks like slugs in WordPress are handled by [`sanitize_title_with_dashes()`](https://developer.wordpress.org/reference/functions/sanitize_title_with_dashes/) according to this [Stack Exchange answer](https://wordpress.stackexchange.com/a/149192).

So I would be inclined to further restrict the non-ascii section to characters that get output from that function as well.

And I think somewhere in the global styles code, we also restrict slugs to not contain consecutive hyphens (`--`) since those would get converted to pipes (`|`) in the `var:` syntax.

jeryj committed:
Return early if no block content

For some reason, the render_block filter runs even on blocks that are not included on the page. So, if you have a duotone filter defined in your theme.json, such as for the core/post-featured-image block, it will run the block filter even if no featured image is set. As a result, the duotonen defined in theme.json per block would get output even if that block wasn't in use. The only reliable check I could find to determine if the block was in use or not was to check the block content. By bailing early if no block content, we avoid adding unused SVG to the output.

jeryj committed:
Make linter happy

scruffian reviewed (approved):
I think this is looking great, and in my testing its working well.

IMO we should merge this and deal with any more improvements in followups, as this PR is already very large.

ajlende commented on lib/class-wp-duotone.php:
```diff
@@ -0,0 +1,150 @@
+ *
+ * @access public
+ */
+class WP_Duotone {
```
Did we figure out if files in the `lib` folder get the `_Gutenberg` suffix. If not, this should have that suffix because we want to be able to make changes in Gutenberg without affecting core by referencing the `_Gutenberg` version.

ajlende commented on lib/class-wp-theme-json-gutenberg.php:
```diff
@@ -3480,4 +3480,39 @@ public function set_spacing_sizes() {
+	 * @param string $slug Slug of the preset.
+	 * @return string CSS variable.
+	 */
+	public static function get_preset_css_var( $path, $slug ) {
+		$duotone_preset_metadata = static::get_preset_metadata_from_path( $path );
+		return static::replace_slug_in_string( $duotone_preset_metadata['css_vars'], $slug );
+	}
```
I think we can get this working without adding this to the public API. The only place the `css_vars` for frontend duotone is being used is within the block supports now, and we already have it hardcoded a couple times there.

oandregal replied:
That'd be lovely. For the record, I saw that this PR was reverted at https://github.com/WordPress/gutenberg/pull/49102

jeryj replied:
This version of the PR did get reverted, and then was cleaned up and merged in this PR:  https://github.com/WordPress/gutenberg/pull/49103

We were able to get the PR merged without adding anything to `lib/class-wp-theme-json-gutenberg.php`
